# Workbook

Use this notebook to complete the exercises throughout the workshop.

#### Table of Contents
- [Section 1 &ndash; Getting Started with Pandas](#Section-1)
- [Section 2 &ndash; Data Wrangling](#Section-2)
- [Section 3 &ndash; Data Visualization](#Section-3)

--- 

### Section 1

#### Exercise 1.1
##### Create a DataFrame by reading in the `2019_Yellow_Taxi_Trip_Data.csv` file. Examine the first 5 rows.

In [2]:
import pandas as pd

In [3]:
taxi_data = pd.read_csv('https://data.cityofnewyork.us/resource/2upf-qytp.csv', nrows=5)
taxi_data

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
0         2  2019-08-02T16:57:13.000  2019-08-02T17:02:58.000   
1         1  2019-08-02T16:54:48.000  2019-08-02T16:58:57.000   
2         2  2019-08-02T16:40:09.000  2019-08-02T17:26:18.000   
3         1  2019-08-02T16:25:44.000  2019-08-02T17:04:45.000   
4         2  2019-08-02T16:47:54.000  2019-08-02T16:57:32.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
0                1           0.97           1                  N   
1                1           0.40           1                  N   
2                1           8.30           1                  N   
3                0           6.70           1                  N   
4                1           1.41           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
0           236            75             2          6.0    1.0      0.5   
1            79            79             4          4.5    3.5      0.5   
2           138           164             1         35.0    1.0      0.5   
3           138            28             2         27.5    1.0      0.5   
4           162           233             1          8.0    1.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        0.00          0.00                    0.3         10.30   
1        0.00          0.00                    0.3          8.80   
2        9.08          6.12                    0.3         54.50   
3        0.00          0.00                    0.3         29.30   
4        2.46          0.00                    0.3         14.76   

   congestion_surcharge  
0                   2.5  
1                   2.5  
2                   2.5  
3                   0.0  
4                   2.5

#### Exercise 1.2
##### Find the dimensions (number of rows and number of columns) in the data.

In [4]:
taxi_data = pd.read_csv('https://data.cityofnewyork.us/resource/2upf-qytp.csv')
taxi_data.shape

(1000, 18)

#### Exercise 1.3
##### Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.

In [5]:
taxi_data[['fare_amount','tip_amount','tolls_amount','total_amount']].describe()

fare_amount   tip_amount  tolls_amount  total_amount
count  1000.000000  1000.000000   1000.000000   1000.000000
mean     12.998900     2.119070      0.220320     18.936840
std      10.355643     2.348863      1.173074     12.300874
min     -10.000000     0.000000      0.000000    -13.800000
25%       7.000000     0.000000      0.000000     11.800000
50%      10.000000     1.955000      0.000000     15.360000
75%      15.000000     3.000000      0.000000     21.247500
max      68.000000    14.000000     12.240000     74.750000

#### Exercise 1.4
##### Isolate the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` for the longest trip by distance (`trip_distance`).

In [6]:
taxi_data[taxi_data.trip_distance==max(taxi_data.trip_distance)][['fare_amount','tip_amount','tolls_amount','total_amount']]



fare_amount  tip_amount  tolls_amount  total_amount
583         68.0         0.0           0.0          69.3

---

### Section 2

#### Exercise 2.1
##### Read in the meteorite data from the `Meteorite_Landings.csv` file, rename the `mass (g)` column to `mass`, and drop all the latitude and longitude columns. Sort the result by mass in descending order.

In [7]:
df = pd.read_csv('https://data.nasa.gov/resource/gh4g-9sfh.csv')
df.rename(columns = {'mass':'mass'}, inplace = True) #this is how you rename a column, but it already was mass, not mass (g)- first the name it is now, then the one you want to change it to
df.sort_values(by = 'mass', ascending = False)

name     id nametype    recclass        mass  fall  \
920   Sikhote-Alin  23593    Valid  Iron, IIAB  23000000.0  Fell   
419          Jilin  12171    Valid          H5   4000000.0  Fell   
29         Allende   2278    Valid         CV3   2000000.0  Fell   
506  Kunya-Urgench  12379    Valid          H5   1100000.0  Fell   
707  Norton County  17922    Valid     Aubrite   1100000.0  Fell   
..             ...    ...      ...         ...         ...   ...   
775             Pê  18179    Valid          L6         NaN  Fell   
787     Pettiswood  18804    Valid  Stone-uncl         NaN  Fell   
793         Picote  18816    Valid  Stone-uncl         NaN  Fell   
869          Sagan  22796    Valid  Stone-uncl         NaN  Fell   
957      Stolzenau  23726    Valid  Stone-uncl         NaN  Fell   

                        year    reclat    reclong             geolocation  
920  1947-01-01T00:00:00.000  46.16000  134.65333      (46.16, 134.65333)  
419  1976-01-01T00:00:00.000  44.05000  126.16667      (44.05, 126.16667)  
29   1969-01-01T00:00:00.000  26.96667 -105.31667  (26.96667, -105.31667)  
506  1998-01-01T00:00:00.000  42.25000   59.20000           (42.25, 59.2)  
707  1948-01-01T00:00:00.000  39.68333  -99.86667   (39.68333, -99.86667)  
..                       ...       ...        ...                     ...  
775  1989-01-01T00:00:00.000  11.33367   -3.54217    (11.33367, -3.54217)  
787  1779-01-01T00:00:00.000  53.53333   -7.33333    (53.53333, -7.33333)  
793  1843-01-01T00:00:00.000  41.36667   -6.23333    (41.36667, -6.23333)  
869  1636-01-01T00:00:00.000  51.53333   14.88333    (51.53333, 14.88333)  
957  1647-01-01T00:00:00.000  52.53333    9.05000        (52.53333, 9.05)  

[1000 rows x 10 columns]

#### Exercise 2.2
##### Using the meteorite data from the `Meteorite_Landings.csv` file, update the `year` column to only contain the year, convert it to a numeric data type, and create a new column indicating whether the meteorite was observed falling before 1970. Set the index to the `id` column and extract all the rows with IDs between 10,036 and 10,040 (inclusive) with `loc[]`.

###### **Hint 1**: Use `year.str.slice()` to grab a substring.

###### **Hint 2**: Make sure to sort the index before using `loc[]` to select the range.

###### **Bonus**: There's a data entry error in the `year` column. Can you find it? (Don't spend too much time on this.)

In [8]:
df['year'] = pd.to_datetime(df['year'], errors = 'coerce')
df['year']= df['year'].dt.year
df

name     id nametype               recclass      mass  fall    year  \
0       Aachen      1    Valid                     L5      21.0  Fell  1880.0   
1       Aarhus      2    Valid                     H6     720.0  Fell  1951.0   
2         Abee      6    Valid                    EH4  107000.0  Fell  1952.0   
3     Acapulco     10    Valid            Acapulcoite    1914.0  Fell  1976.0   
4      Achiras    370    Valid                     L6     780.0  Fell  1902.0   
..         ...    ...      ...                    ...       ...   ...     ...   
995   Tirupati  24009    Valid                     H6     230.0  Fell  1934.0   
996    Tissint  54823    Valid  Martian (shergottite)    7000.0  Fell  2011.0   
997      Tjabe  24011    Valid                     H6   20000.0  Fell  1869.0   
998   Tjerebon  24012    Valid                     L5   16500.0  Fell  1922.0   
999  Tomakovka  24019    Valid                    LL6     600.0  Fell  1905.0   

       reclat    reclong            geolocation  
0    50.77500    6.08333      (50.775, 6.08333)  
1    56.18333   10.23333   (56.18333, 10.23333)  
2    54.21667 -113.00000     (54.21667, -113.0)  
3    16.88333  -99.90000      (16.88333, -99.9)  
4   -33.16667  -64.95000    (-33.16667, -64.95)  
..        ...        ...                    ...  
995  13.63333   79.41667   (13.63333, 79.41667)  
996  29.48195   -7.61123   (29.48195, -7.61123)  
997  -7.08333  111.53333  (-7.08333, 111.53333)  
998  -6.66667  106.58333  (-6.66667, 106.58333)  
999  47.85000   34.76667      (47.85, 34.76667)  

[1000 rows x 10 columns]

In [10]:
import numpy as np
df['year']=df.year.astype(str)
df['fallen_before_1970'] = np.where(df['year']<= '1970', True, False)

df

name     id nametype               recclass      mass  fall    year  \
0       Aachen      1    Valid                     L5      21.0  Fell  1880.0   
1       Aarhus      2    Valid                     H6     720.0  Fell  1951.0   
2         Abee      6    Valid                    EH4  107000.0  Fell  1952.0   
3     Acapulco     10    Valid            Acapulcoite    1914.0  Fell  1976.0   
4      Achiras    370    Valid                     L6     780.0  Fell  1902.0   
..         ...    ...      ...                    ...       ...   ...     ...   
995   Tirupati  24009    Valid                     H6     230.0  Fell  1934.0   
996    Tissint  54823    Valid  Martian (shergottite)    7000.0  Fell  2011.0   
997      Tjabe  24011    Valid                     H6   20000.0  Fell  1869.0   
998   Tjerebon  24012    Valid                     L5   16500.0  Fell  1922.0   
999  Tomakovka  24019    Valid                    LL6     600.0  Fell  1905.0   

       reclat    reclong            geolocation  fallen_before_1970  
0    50.77500    6.08333      (50.775, 6.08333)                True  
1    56.18333   10.23333   (56.18333, 10.23333)                True  
2    54.21667 -113.00000     (54.21667, -113.0)                True  
3    16.88333  -99.90000      (16.88333, -99.9)               False  
4   -33.16667  -64.95000    (-33.16667, -64.95)                True  
..        ...        ...                    ...                 ...  
995  13.63333   79.41667   (13.63333, 79.41667)                True  
996  29.48195   -7.61123   (29.48195, -7.61123)               False  
997  -7.08333  111.53333  (-7.08333, 111.53333)                True  
998  -6.66667  106.58333  (-6.66667, 106.58333)                True  
999  47.85000   34.76667      (47.85, 34.76667)                True  

[1000 rows x 11 columns]

In [18]:
df = df.loc[(df['id'] > 10036) & (df['id'] <= 10040)] 

df


name     id nametype recclass      mass  fall    year    reclat  \
282      Enshi  10038    Valid       H5    8000.0  Fell  1974.0  30.30000   
283  Ensisheim  10039    Valid      LL6  127000.0  Fell     nan  47.86667   

     reclong       geolocation  fallen_before_1970  
282   109.50     (30.3, 109.5)               False  
283     7.35  (47.86667, 7.35)               False

#### Exercise 2.3
##### Using the meteorite data from the `Meteorite_Landings.csv` file, create a pivot table that shows both the number of meteorites and the 95th percentile of meteorite mass for those that were found versus observed falling per year from 2005 through 2009 (inclusive). Hint: Be sure to convert the `year` column to a number as we did in the previous exercise.

#### Exercise 2.4
##### Using the meteorite data from the `Meteorite_Landings.csv` file, compare summary statistics of the mass column for the meteorites that were found versus observed falling.

#### Exercise 2.5
##### Using the taxi trip data in the `2019_Yellow_Taxi_Trip_Data.csv` file, resample the data to an hourly frequency based on the dropoff time. Calculate the total `trip_distance`, `fare_amount`, `tolls_amount`, and `tip_amount`, then find the 5 hours with the most tips.

--- 

### Section 3

#### Exercise 3.1
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create box plots for traveler throughput for each year in the data. Hint: Pass `kind='box'` into the `plot()` method to generate box plots.

#### Exercise 3.2
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create a heatmap that shows the 2019 TSA median traveler throughput by day of week and month.

#### Exercise 3.3
##### Annotate the medians in the box plot from *[Exercise 3.1](#Exercise-3.1)*. Hint: The `x` coordinates will be 1, 2, and 3 for 2019, 2020, and 2021, respectively. Alternatively, to avoid hardcoding values, you can use the `Axes.get_xticklabels()` method, in which case you should look at the [documentation](https://matplotlib.org/stable/api/text_api.html) for the `Text` class.